src/libraries/Common/src/System/Security/Cryptography/X509Certificates/X509CertificateLoader.Pkcs12.cs

```diff
@@ -0,0 +1,909 @@
+
+                        if (bag.BagAttributes is not null)
+                        {
+                            FilterAttributes(
+                                loaderLimits,
+                                ref bag,
+                                static (limits, oid) => limits.PreserveUnknownAttributes || oid == Oids.LocalKeyId);
+                        }
+
+                        bagState.AddCert(bag);
+                    }
+                }
+                else if (bag.BagId == Oids.Pkcs12KeyBag || bag.BagId == Oids.Pkcs12ShroudedKeyBag)
+                {
+                    if (bagState.KeyCount >= loaderLimits.MaxKeys)
+                    {
+                        throw new Pkcs12LoadLimitExceededException(nameof(Pkcs12LoaderLimits.MaxKeys));
+                    }
+
+                    if (bag.BagId == Oids.Pkcs12ShroudedKeyBag)
+                    {
+                        EncryptedPrivateKeyInfoAsn epki = EncryptedPrivateKeyInfoAsn.Decode(
+                            bag.BagValue,
+                            AsnEncodingRules.BER);
+
+                        int kdfCount = GetKdfCount(epki.EncryptionAlgorithm);
+
+                        if (kdfCount > loaderLimits.IndividualKdfIterationLimit || kdfCount > workRemaining)
+                        {
+                            throw new Pkcs12LoadLimitExceededException(
+                                kdfCount > loaderLimits.IndividualKdfIterationLimit ?
+                                    nameof(Pkcs12LoaderLimits.IndividualKdfIterationLimit) :
+                                    nameof(Pkcs12LoaderLimits.TotalKdfIterationLimit));
+                        }
+
+                        if (workRemaining.HasValue)
+                        {
+                            workRemaining -= kdfCount;
+                        }
+                    }
+
+                    if (bag.BagAttributes is not null)
+                    {
+                        FilterAttributes(
+                            loaderLimits,
+                            ref bag,
+                            static (limits, attrType) =>
+                                attrType switch
+                                {
+                                    Oids.LocalKeyId => true,
+                                    "1.2.840.113549.1.9.20" => limits.PreserveKeyName,
+                                    "1.3.6.1.4.1.311.17.1" => limits.PreserveStorageProvider,
+                                    _ => limits.PreserveUnknownAttributes,
+                                });
+                    }
+
+                    bagState.AddKey(bag);
+                }
+            }
+        }
+
+        private static void FilterAttributes(
+            Pkcs12LoaderLimits loaderLimits,
+            ref SafeBagAsn bag,
+            Func<Pkcs12LoaderLimits, string, bool> filter)
+        {
+            if (bag.BagAttributes is not null)
+            {
+                // Should this dedup/fail-on-dup?
```

Comment:
I would vote fail, though nothing in the RFC seems to indicate that is the “right” behavior since the RFC doesn’t really say much of anything about it other than that it is a SET OF.

Reply:
While mathematical sets can't have duplicates, ASN.1 `SET OF` is just an "unordered" sequence (aside from the CER/DER ordering requirements).  I don't think there's anything in the structure that inherently says "no dupes", but since dupes are probably always malicious fail-on-dupe is nice... other than the cost of detecting it.

Reply:
I guess if we are re-writing the attributes we can pick "first wins". My more immediate concern was platforms having differences of opinions on dupes if we sent through as-is.